presto-spi/src/main/java/com/facebook/presto/spi/function/LambdaDescriptor.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.function;
+
+import com.facebook.presto.common.Subfield;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+import static java.util.Objects.requireNonNull;
+
+public class LambdaDescriptor
+{
+    /**
+     * Index of the argument in CallExpression, which represents this function, that contains the functional parameter.
+     * Value -1 indicates that the function does not accept the functional argument.
+     */
+    final int callArgumentIndex;
+    /**
+     * The mapping of the index of lambda argument to the index of the argument of the CallExpression that this argument relates to.
+     * <p>
+     * Example: ZIP_WITH(array_column1, array_column2, (x,y) -> x.quantity * y.price).
+     * The purpose of this mapping is to correctly attribute that subfield `quantity` relates to the `array_column1`, while subfield `price` relates to the `array_column2`.
+     * Here, the lambda argument 0 with the name x relates to the argument 0 in CallExpression that represent ZIP_WITH function, while the lambda argument 1 corresponds
+     * to the argument 1 of the CallExpression. Therefore, the mapping for this example must be as follows:
+     * { 0 => 0, 1 => 1 }.
+     */
+    final Map<Integer, Integer> lambdaArgumentToCallArgumentIndexMap;
+
+    /**
+     * Contains the transformation function that needs to be applied to the lambda subfields of this function.
```

Comment:
Can you explain more what this is.

Reply:
collected subfields from lambda do not correspond 1-to-1 to the input array or map argument. For instance, 
```
CREATE TABLE my_table( array_data ARRAY( ROW( id BIGINT, data VARCHAR ) ) )
SELECT any_match( array_data, r -> r.id > 100 ) FROM my_table
```
We collected the `r.id` from the lambda that does not correspond 1-to-1 to the input array `array_data`. We need a path transformation function to transform it from `id` to `[*].id` (prepend `allSubfields` path element). `array_data` is not a part of the path and will be added once we reach the input (e.g. variable reference expression). 

`TRANSFORM`, `ZIP_WITH`,  `TRANSFORM_VALUES`, and `REDUCE` functions add all the complexity requiring it to be very generic so that we can later extend the level of support of those functions.